### PR TITLE
Add Ready to Pay API

### PR DESCRIPTION
**Pull request by justnero**

### CODE REVIEW

**Comment by justnero** [comment]
php please review

**Comment by justnero** [comment]
php please review

**Comment by justnero** [comment]
php please review

File: openapi/components/schemas/ReadyToPay/ReadyToPay.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+type: object
+required:
+  - websiteId
+  - browserData
+properties:
+  websiteId:
+    description: The website identifier string
+    type: string
```
Comment (adamaltman) [inline openapi/components/schemas/ReadyToPay/ReadyToPay.yaml:8]: Is this required? You already have the `allOf`.

File: openapi/paths/storefront/ready-to-pay.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  summary: Ready to Pay
+  operationId: StorefrontPostReadyToPay
+  security:
+    - PublishableApiKey: []
+  description: |
+    Get available payment methods for a specific transaction or a purchase.
+
+    The payment methods order shown to a customer **MUST** be the same as the order in the response.
+
+    If Ready to Pay is used with Transaction-scoped JWT all fields will be ignored except of the `browserData`.
```
Comment (adamaltman) [inline openapi/paths/storefront/ready-to-pay.yaml:13]:
```suggestion
    If Ready to Pay is used with Transaction-scoped JWT all fields will be ignored except the `browserData`.
```

Comment (adamaltman) [reply openapi/paths/storefront/ready-to-pay.yaml:13]: This is because it takes the data from the transaction?

File: openapi/paths/storefront/ready-to-pay.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  tags:
+    - Purchase
+  summary: Ready to Pay
+  operationId: StorefrontPostReadyToPay
+  security:
+    - PublishableApiKey: []
+  description: |
+    Get available payment methods for a specific transaction or a purchase.
+
+    The payment methods order shown to a customer **MUST** be the same as the order in the response.
```
Comment (adamaltman) [inline openapi/paths/storefront/ready-to-pay.yaml:11]:
```suggestion
    The payment methods order shown to a customer **SHOULD** be the same as the order in the response.
```
I think they won't get in trouble if it isn't (typically)... so I think it's a SHOULD.

File: openapi/paths/ready-to-pay.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+post:
+  tags:
+    - Transactions
+  summary: Ready to Pay
+  operationId: PostReadyToPay
+  description: |
+    Get available payment methods for a specific transaction or a purchase.
+
+    The payment methods order shown to a customer **MUST** be the same as the order in the response.
```
Comment (adamaltman) [inline openapi/paths/ready-to-pay.yaml:9]:
```suggestion
    The payment methods order shown to a customer **SHOULD** be the same as the order in the response.
```

File: openapi/components/schemas/ReadyToPay/Methods/ReadyToPayPaymentCardMethod.yaml [hunk 5]
```diff
@@ -0,0 +1,43 @@
+type: object
+title: Payment Card
+required:
+  - method
+properties:
+  method:
+    type: string
+    description: The payment method.
```
Comment (vykhrystiuk) [inline openapi/components/schemas/ReadyToPay/Methods/ReadyToPayPaymentCardMethod.yaml:8]: Do we need a period at the and here? Afaik, we need it in the description (if it's a sentence)

Otherwise, it's different [here](https://github.com/Rebilly/api-definitions/pull/197/files#diff-bb174cdf38ce5f40c4c35d0934e26984R7). Please check other places ([here](https://github.com/Rebilly/api-definitions/pull/197/files#diff-a6f46f5dbbafdfd73a347f3099bab175R34), etc.) as well.

File: openapi/paths/storefront/ready-to-pay.yaml (hunk 2 above)
Comment (justnero) [reply openapi/paths/storefront/ready-to-pay.yaml:13]: Yes

**Review by iamdevice — DISMISSED** [review]

**Review by ashkarpetin — DISMISSED** [review]

File: openapi/components/schemas/ReadyToPay/ReadyToPayMethods.yaml [hunk 6]
```diff
@@ -21,12 +22,12 @@ example:
...
-      values:
-        - "411111"
-        - "444433"
+    filters:
+      - "brand:Visa;bin:411111,444433"
   - method: ach
     feature: null
+    filters: []
   - method: paypal
+    filters: [ ]
```
Comment (iamdevice) [inline openapi/components/schemas/ReadyToPay/ReadyToPayMethods.yaml:31]: Maybe you wanted to do next?
```suggestion
    filters: []
```
also to Skrill

File: openapi/components/schemas/EventType.yaml [hunk 7]
```diff
@@ -4,15 +4,37 @@ readOnly: true
 enum:
   - dispute-created
   - gateway-account-requested
-  - invoice-paid
-  - invoice-past-due
-  - payment-card-expired
-  - risk-score-changed
-  - subscription-activated
+  - transaction-processed
```
Comment (XaosSintez) [inline openapi/components/schemas/EventType.yaml:7]: As i see, items were sorted alphabetically. Now it looks like a mess. Any reason of such changes?

Comment (justnero) [reply openapi/components/schemas/EventType.yaml]: My bad, forgot to sort them out

File: openapi/components/schemas/ReadyToPay/Methods/ReadyToPayAchMethod.yaml [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  method:
+    type: string
+    description: The payment method
+    enum:
+      - ach
+  feature:
+    type: object
+    nullable: true
+    description: |
+      The specifc feature (eg. digital wallet or a processor) of this method. If method doesn't have any features – will be null.
```
Comment (svensono) [inline openapi/components/schemas/ReadyToPay/Methods/ReadyToPayAchMethod.yaml:15]:
```suggestion
      The specific feature (eg. digital wallet or a processor) of this method. If method doesn't have any features – will be null.
```

File: openapi/components/schemas/ReadyToPay/Methods/ReadyToPayPaymentCardMethod.yaml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  method:
+    type: string
+    description: The payment method
+    enum:
+      - payment-card
+  feature:
+    type: object
+    nullable: true
+    description: |
+      The specifc feature (eg. digital wallet or a processor) of this method. If method doesn't have any features – will be null.
```
Comment (svensono) [inline openapi/components/schemas/ReadyToPay/Methods/ReadyToPayPaymentCardMethod.yaml:15]:
```suggestion
      The specific feature (eg. digital wallet or a processor) of this method. If method doesn't have any features – will be null.
```

File: openapi/components/schemas/ReadyToPay/Methods/ReadyToPayPaymentCardMethod.yaml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    oneOf:
+      - $ref: ../Features/PaymentCardDigitalWalletFeature.yaml
+  brands:
+    type: array
+    description: The list of supported brands.
+    minItems: 1
+    items:
+      $ref: ../../PaymentCards/PaymentCardBrand.yaml
+  filters:
+    type: array
```
Comment (XaosSintez) [inline openapi/components/schemas/ReadyToPay/Methods/ReadyToPayPaymentCardMethod.yaml:25]: Our filters common format is string

File: openapi/components/schemas/ReadyToPay/ReadyToPayAmount.yaml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+type: object
+title: 'With amount'
+required:
+  - currency
+  - amount
+properties:
+  currency:
+    allOf:
+      - $ref: ../CurrencyCode.yaml
```
Comment (XaosSintez) [inline openapi/components/schemas/ReadyToPay/ReadyToPayAmount.yaml:9]: Can't we use just `$ref` here?
```
currency:
  $ref: ../CurrencyCode.yaml
```

File: openapi/paths/storefront/ready-to-pay.yaml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+    The payment methods order shown to a customer **SHOULD** be the same as the order in the response.
+
+    If Ready to Pay is used with Transaction-scoped JWT all fields will be ignored except the `browserData`.
+  requestBody:
+    content:
+      application/json:
+        schema:
+          $ref: ../../components/schemas/ReadyToPay/ReadyToPay.yaml
+
```
Comment (svensono) [inline openapi/paths/storefront/ready-to-pay.yaml:19]:
```suggestion

```

File: openapi/paths/ready-to-pay.yaml [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  description: |
+    Get available payment methods for a specific transaction or a purchase.
+
+    The payment methods order shown to a customer **SHOULD** be the same as the order in the response.
+  requestBody:
+    content:
+      application/json:
+        schema:
+          $ref: ../components/schemas/ReadyToPay/CoreReadyToPay.yaml
+
```
Comment (svensono) [inline openapi/paths/ready-to-pay.yaml:15]:
```suggestion

```

File: openapi/components/schemas/Rules/Actions/adjust-ready-to-pay.yaml [hunk 14]
```diff
@@ -0,0 +1,107 @@
+allOf:
+  - $ref: ../RuleAction.yaml
+  - type: object
+    description: Add risk score
+    properties:
+      gatewayAccountsMode:
+        type: string
+        description: |
+          The global gateway accounts mode that will be applied at the beginning of the action processing.
```
Comment (svensono) [inline openapi/components/schemas/Rules/Actions/adjust-ready-to-pay.yaml:9]:
```suggestion
          The global gateway account mode that will be applied at the beginning of the action processing.
```
Not entirely sure about this

File: openapi/components/schemas/Rules/Actions/adjust-ready-to-pay.yaml [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+          - `allow-all` will mark all available gateway accounts as allowed if they were denied by any previous actions.
+          - `deny-all` will mark all available gateway accounts as denied if they were allowed by any previous actions.
+          - `manual` will not apply any global allow/deny rule.
+        enum:
+          - manual
+          - allow-all
+          - deny-all
+      gatewayAccounts:
+        type: array
+        description: The individual list of gateway accounts adjustments
```
Comment (svensono) [inline openapi/components/schemas/Rules/Actions/adjust-ready-to-pay.yaml:20]:
```suggestion
        description: The individual list of gateway account adjustments
```

File: openapi/components/schemas/Rules/Actions/adjust-ready-to-pay.yaml [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+          - `allow-all` will mark all available payment methods as allowed if they were denied by any previous actions.
+          - `deny-all` will mark all available payment methods as denied if they were allowed by any previous actions.
+          - `manual` will not apply any global allow/deny rule.
+        enum:
+          - manual
+          - allow-all
+          - deny-all
+      paymentMethods:
+        type: array
+        description: The individual list of gateway accounts adjustments
```
Comment (svensono) [inline openapi/components/schemas/Rules/Actions/adjust-ready-to-pay.yaml:47]:
```suggestion
        description: The individual list of gateway account adjustments
```

File: openapi/components/schemas/Rules/Actions/adjust-ready-to-pay.yaml [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+            paymentMethod:
+              description: The payment method name
+            feature:
+              description: |
+                The specific payment method feature like `Google Pay`, `Apple Pay` or `Plaid`.
+
+                If set to `null` – applies to the payment method itself.
+            mode:
+              type: string
+              description: The individual payment metod mode
```
Comment (svensono) [inline openapi/components/schemas/Rules/Actions/adjust-ready-to-pay.yaml:97]:
```suggestion
              description: The individual payment method mode
```

**Comment by justnero** [comment]
php please re-review

File: openapi/components/schemas/ReadyToPay/ReadyToPayItems.yaml [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+    items:
+      type: object
+      required:
+        - planId
+      properties:
+        planId:
+          description: The plan identifier string
+          allOf:
+            - $ref: ../ResourceId.yaml
+        quantity:
```
Comment (XaosSintez) [inline openapi/components/schemas/ReadyToPay/ReadyToPayItems.yaml:18]: I think it has default value?

File: openapi/components/schemas/Rules/Actions/adjust-ready-to-pay.yaml [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+            mode:
+              type: string
+              description: The individual payment method mode
+              enum:
+                - allow
+                - deny
+            weight:
+              type: integer
+              description: |
+                The weight is used to order (decending) the payment method in the final response. It can be negative.
```
Comment (XaosSintez) [inline openapi/components/schemas/Rules/Actions/adjust-ready-to-pay.yaml:104]:
```suggestion
                The weight is used to order (descending) the payment method in the final response. It can be negative.
```

File: openapi/components/schemas/ReadyToPay/ReadyToPayItems.yaml (hunk 18 above)
Comment (justnero) [reply openapi/components/schemas/ReadyToPay/ReadyToPayItems.yaml:19]: No, it should be required, fixed.

**Review by XaosSintez — DISMISSED** [review]
Looks ok to me

File: openapi/components/schemas/EventType.yaml [hunk 20]
```diff
@@ -2,16 +2,38 @@ type: string
 description: Rebilly event type
 readOnly: true
 enum:
+  - account-password-reset-requested
```
Comment (iamdevice) [inline openapi/components/schemas/EventType.yaml:5]: I think we should make these changes in different PRs. I'd like to see only `ready-to-pay-requested` event in that PR.

**Review by iamdevice — APPROVED** [review]

**Review by subenson — APPROVED** [review]
